### ISSUE TITLE
MediaSDK 2018R1 vs latest git (9890dff) output file size

### ISSUE THREAD

**Issue by ViGill** [Intel-Media-SDK/MediaSDK#1992]
Hi,

we are considering moving for MediaSDK 2018R1 to the latest git code.
The current package I am using is MediaServerStudioEssentials2018R1.
So far so good, but I noticed something very strange. The resulting file size is much lower with the latest MSDK release. Quality is a bit lower as well... 
I use the same FFmpeg binary in both cases, with same patched kernel (4.14).

The transcoding command line is:
ffmpeg -v debug  -hwaccel qsv -init_hw_device qsv=myqsv -filter_hw_device myqsv -qsv_device /dev/dri/renderD128 -c:v h264_qsv  -i [file]  -an  -vcodec h264_qsv -preset slow -me_range 8 -map_chapters -1 -profile:v high -level:v 31 -max_muxing_queue_size 10000 -g 72 -aspect 16/9 -look_ahead 1 -look_ahead_depth 30 -b:v 2000000  -vsync 1 -r 24.000000 -vf [main]hwupload,vpp_qsv=w=1280:h=720[out]  -threads 4  -y -stats /tmp/out.h264

You can find attached both logs with -v debug enable.
[msdk_2018R1.log](https://github.com/Intel-Media-SDK/MediaSDK/files/4280743/msdk_2018R1.log)
[msdk_9890dff.log](https://github.com/Intel-Media-SDK/MediaSDK/files/4280744/msdk_9890dff.log)

With MediaServerStudioEssentials2018R1: file size is 36700kB
With 9890dff: file size is 33898kB

Thanks!

**Comment by ViGill** [comment]
[msdk_2018R1.log](https://github.com/Intel-Media-SDK/MediaSDK/files/4280750/msdk_2018R1.log)
[msdk_9890dff.log](https://github.com/Intel-Media-SDK/MediaSDK/files/4280751/msdk_9890dff.log)

**Comment by dmitryermilov** [comment]
Hi @ViGill ,

It's 2 years difference between MSS and latest gitHub. And difference is not only on MediaSDK side but on driver level as well. I mean it's impossible to say what caused behavior change without deep investigation. I don't think we can jump into it without confirmation it's a obvious regression seen on a range of different video streams. E.g. could you check BD rate to confirm?

**Comment by ViGill** [comment]
Hi @dmitryermilov , for the answer!
Yes it can be the driver as well.. lots of changes since then. What do you mean by BD?

Also, i just tested on kernel 4.15.0-58 and it works! Though, the transcoding speed goes from around 500FPS to 300FPS when transcoding from 1920x1080 to 512x288p, h264 compared to kernel 4.14.20 with 2018R2 patches ... any idea why ?

thanks

**Comment by dmitryermilov** [comment]
>What do you mean by BD?

Please look at https://streaminglearningcenter.com/encoding/compute-bd-rate-functions.html . It has an overview. 

As for performance drop, please take a look at https://github.com/Intel-Media-SDK/MediaSDK/issues/95 . Could you try with 5.3 ?

**Comment by ViGill** [comment]
Hi,

sure I will try 5.3!
I will look at BD rate, thanks for the link.

**Comment by ViGill** [comment]
uname -a
Linux worker-068.adm.dc3.dailymotion.com 5.3.0-40-generic #32~18.04.1-Ubuntu SMP Mon Feb 3 14:05:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

rame= 3600 fps=488 q=10.0 Lsize=    7874kB time=00:02:30.07 bitrate= 429.8kbits/s speed=20.3x

well done! 5.3 does indeed fix it!
Thanks

**Comment by dmitryermilov** [comment]
Good! You're welcome

**Comment by dmitryermilov** [comment]
Hi @ViGill ,

Any news about VQ?

**Comment by ViGill** [comment]
Hi  @dmitryermilov,
not yet but I am still not convinced by the new drivers :)
Let me explain. I am not running the following packages:
FFmpeg 4.2.2
intel-gmmlib-19.4.1
intel-media-19.4.0
intel-mediasdk-19.4.0
libdrm-2.4.100
libva-2.6.1
Linux Kernel vanilla 5.3

I use the following command to transcode:

ffmpeg -hwaccel qsv -init_hw_device qsv=myqsv -filter_hw_device myqsv -qsv_device /dev/dri/renderD128 -c:v h264_qsv -autorotate 0 -i ./Pacific_Rim_Trailer_2_5.1-1080p-HDTN.mp4 -metadata:s:v:0 rotate=0 -an -vcodec h264_qsv -preset medium -me_range 8 -map_metadata -1 -map_chapters -1 -profile:v main -level:v 22 -max_muxing_queue_size 10000 -g 72 -aspect 16/9 -look_ahead 1 -look_ahead_depth 30 -b:v 392000 -winmaxavg 440 -winsize 72 -vsync 1 -r 24.000000 -vf [main]hwupload,vpp_qsv=w=512:h=288[out] -f mp4 -movflags faststart -threads 4 -fs 40000000000 -y -stats out.mp4

the resulting file has the following video track:

Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : Main@L2.2
Format settings, CABAC                   : Yes
Format settings, ReFrames                : 3 frames
Format settings, GOP                     : M=4, N=72
Codec ID                                 : avc1
Codec ID/Info                            : Advanced Video Coding
Duration                                 : 2 min 30 s
Bit rate                                 : 313 kb/s
Width                                    : 512 pixels
Height                                   : 288 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Constant
Frame rate                               : 24.000 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Bits/(Pixel*Frame)                       : 0.088
Stream size                              : 5.60 MiB (99%)
Encoded date                             : UTC 1904-01-01 00:00:00
Tagged date                              : UTC 1904-01-01 00:00:00

As you can see, the resulting bitrate is much lower than the one I requested (313 kb/s instead of 392).
I did not have this behavior 2 years ago.
So appart for quality considerations, I think there is something wrong ...
Could you please help me figure out what ?

Thanks a lot

**Comment by dmitryermilov** [comment]
Actual bitrate is 25% higher than requested. I agree that it's poor accuracy.  Could you prepare a reproducer with sample_encode and share an input YUV file?

**Comment by ViGill** [comment]
Hi, you mean 25% lower. You mean reproducing with the exact same encoding paramerters ? This is challenging but I can try :)

**Comment by dmitryermilov** [comment]
Ahh, right, lower.
I asked about reproducer to simplify it for developers. Thanks!

**Comment by ViGill** [comment]
Yes of course I understand, no problem I'm on it! Thanks

**Comment by ViGill** [comment]
Hi,

after more investigations, it looks like it comes from lookahead...
I achieve the same result with the sample_encode app.

The fifo.yuv fifo is fed by the 2m30s file this way:

```
ffmpeg -y -i  ./Pacific_Rim_Trailer_2_5.1-1080p-HDTN.mp4  -c:v rawvideo -pix_fmt yuv420p -f rawvideo /tmp/fifo.yuv```

With lookahead:
```
./sample_encode h264 -i /tmp/fifo.yuv -o encode.h264 -w 1920 -h 1038 -f 24 -b 392 -MaxKbps 440 -u balanced -g 72 -la -hw -dstw 512 -dsth 288  -num_slice 1 -x 3 -CodecLevel 22 -lad 10 -ws 72 -wb 440

encode.h264: 6028264 bytes -> 313 Kbit/s 
```

Without lookahead
```
./sample_encode h264 -i /tmp/fifo.yuv -o encode_nola.h264 -w 1920 -h 1038 -f 24 -b 392 -MaxKbps 440 -u balanced -g 72 -hw -dstw 512 -dsth 288  -num_slice 1 -x 3 -CodecLevel 22 -ws 72 -wb 440

encode_nola.h264: 7013610 bytes  -> 374Kbit/s
```

**Comment by ViGill** [comment]
File is here: http://dl.free.fr/svE0c5InL

**Comment by dmitryermilov** [comment]
@ogolikov , can you please comment?

**Comment by dmitryermilov** [comment]
@ViGill , please try set minQP = 1

**Comment by ViGill** [comment]
Hi @dmitryermilov , I think I found what I was doing wrong ...
-wb (winmaxavg) value should be higher to cope with video with big bitrate variation
- look_ahead_downsampling set to off also helps having better results

Using wb as 2*video_bitrate works very well, I now get output bitrates that are very close to what I request 

Where could I upload the file ?

Thanks for you great support!

**Comment by dmitryermilov** [comment]
You're welcome!
> look_ahead_downsampling set to off also helps having better results

Good. But please note it may affect performance because motion estimation on "full sized" surfaces will be slower than on downscaled surfaces.

> wb (winmaxavg) value should be higher to cope with video with big bitrate variation

@ogolikov , can you please comment?

Perhaps, google drive? Or you can send your email to dmitry.ermilov@intel.com, and I'll send an invitation to a shared syncplicity folder .

**Comment by dmitryermilov** [comment]
Hi @ViGill ,
We double checked. Basically Look Ahead behaves the right way in comparison to regular hardware BRC. LookAhead BRC tries to follow sliding window BRC condition (defined by winmaxavg). That's why it doesn't keep target bitrate. 
HW BRC looks like ignore sliding window condition. 

So if "wb as 2*video_bitrate" works for you, let's consider the issue addressed. Okay?

**Comment by ViGill** [comment]
Hi @dmitryermilov, thanks for your comment and sorry for the lack of answer the past 12 days, to be honest I did not see the mail notification of your answer :)

Anyway,  I moved forward and deployed this whole new software release on half our encoding farm (our farm is composed of around 180 m710x cartridges).
So it means:
- 90 cartridges running latest MSDK/ffmpeg 4.2.2/Kernel 5.3.0
- 90 cartridges running MSDK 2018R2/ffmpeg 3.4.1/Kernel 1.14 patched

=> After a week of run, and hundred of thousands of encodings, we noticed that the farm running the new  SW is still slower, by around 25/30%. It means that the farm with older SW transcodes around 30% more videos.

So I tried to rollback the kernel to 4.14 on half the farm with the new SDK, which means that have on this farm, 45 cartridges with kernel 4.14 patched and 45 with kernel 5.3.0, using the same MSDK (latest one) and ffmpeg 4.2.2.

=> the 45 cartridges with the older kernel (4.14) are transcoding 30% more videos than the one with kernel 5.3.0.

So it looks like kernel 5.3.0 is still not as good as old kernel 4.14 with patches...
I tried to encode a single file on each config, and the 5.3.0 kernel is only about 5% slower.
So maybe it becomes a problem when doing multiple concurrent transcodings ?

Thanks,
Gilles
